pkg/controller/attach_detach.go

```diff
@@ -120,7 +131,20 @@ func (r *ReconcileAttachDetach) Reconcile(ctx context.Context, request reconcile
 		}
 		// attachment request
 	} else if azVolumeAttachment.Status.Detail == nil {
+		r.SharedState.nodeDiskAvailabilityMapLock.Lock()
+		defer r.SharedState.nodeDiskAvailabilityMapLock.Unlock()
+		// azVolumeAttachment state is nil which means it's a brand new attachment request instead of a retry
+		// the node's capacity of disk attachment should be decreased by 1
+		remainingCapacity, nodeExists := r.SharedState.nodeDiskAvailabilityMap.Load(azVolumeAttachment.Spec.NodeName)
+		if nodeExists && remainingCapacity != nil && remainingCapacity.(int) > 0 {
+			r.SharedState.nodeDiskAvailabilityMap.Store(azVolumeAttachment.Spec.NodeName, remainingCapacity.(int)-1)
+		} else {
+			err := status.Errorf(codes.Aborted, "node (%s) is unavailable for disk attachment", azVolumeAttachment.Spec.NodeName)
+			return reconcileReturnOnError(ctx, azVolumeAttachment, "attach", err, r.retryInfo)
+		}
```

Comment:
I believe it would be better to perform the allocation in `triggerAttach` rather than directly in the reconciler.

Reply:
Done